Invalid suggestion for a reference of iterator

### Code

```Rust
fn main() {
    let fields = vec!["a", "b", "c"];
    let variant = vec!["a", "b", "c"];

    for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
        eprintln!("{} {}", src, dest);
    }
}
```


### Current output

```Shell
error[E0277]: `&std::slice::Iter<'_, &str>` is not an iterator
  --> ./p/iter.rs:5:54
   |
5  |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
   |                        --------------                ^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, &str>` is not an iterator
   |                        |
   |                        required by a bound introduced by this call
   |
   = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, &str>`, which is required by `&std::slice::Iter<'_, &str>: IntoIterator`
   = note: required for `&std::slice::Iter<'_, &str>` to implement `IntoIterator`
note: required by a bound in `std::iter::zip`
  --> /Users/yukang/rust/library/core/src/iter/adapters/zip.rs:70:8
   |
67 | pub fn zip<A, B>(a: A, b: B) -> Zip<A::IntoIter, B::IntoIter>
   |        --- required by a bound in this function
...
70 |     B: IntoIterator,
   |        ^^^^^^^^^^^^ required by this bound in `zip`
help: consider dereferencing here
   |
5  |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter()) {
   |                                                      +
help: consider removing the leading `&`-reference
   |
5  -     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
5  +     for (src, dest) in std::iter::zip(fields.iter(), variant.iter()) {
   |
help: consider changing this borrow's mutability
   |
5  |     for (src, dest) in std::iter::zip(fields.iter(), &mut variant.iter()) {
   |                                                      ~~~~

error[E0277]: `&std::slice::Iter<'_, &str>` is not an iterator
 --> ./p/iter.rs:5:24
  |
5 |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, &str>` is not an iterator
  |
  = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, &str>`, which is required by `Zip<std::slice::Iter<'_, &str>, &std::slice::Iter<'_, &str>>: IntoIterator`
  = help: the trait `Iterator` is implemented for `std::slice::Iter<'a, T>`
  = note: `Iterator` is implemented for `&mut std::slice::Iter<'_, &str>`, but not for `&std::slice::Iter<'_, &str>`
  = note: required for `Zip<std::slice::Iter<'_, &str>, &std::slice::Iter<'_, &str>>` to implement `Iterator`
  = note: required for `Zip<std::slice::Iter<'_, &str>, &std::slice::Iter<'_, &str>>` to implement `IntoIterator`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
```


### Desired output

```Shell
Remove the first suggestion:


help: consider dereferencing here
   |
5  |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter()) {
   |                                                      +
```


### Rationale and extra context

Suggest `dereferencing` is not right here, if we use `*&variant`, the code is still not right, and continue to suggest:

```console
help: consider cloning the value if the performance cost is acceptable
  |
5 -     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter()) {
5 +     for (src, dest) in std::iter::zip(fields.iter(), variant.iter().clone()) {
  |
```

And add more `.clone()` it will continue to suggest add `.clone()` in the next compiling, this is another issue need to be fixed.

```console
help: consider cloning the value if the performance cost is acceptable
  |
5 -     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter().clone()) {
5 +     for (src, dest) in std::iter::zip(fields.iter(), variant.iter().clone().clone()) {
```

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.80.0-nightly (8c127df75 2024-05-16)
binary: rustc
commit-hash: 8c127df75fde3d5ad8ef9af664962a7676288b52
commit-date: 2024-05-16
host: aarch64-apple-darwin
release: 1.80.0-nightly
LLVM version: 18.1.4
```


### Anything else?

_No response_

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -466,7 +466,6 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             && let Some(arg_ty) = typeck_results.expr_ty_adjusted_opt(expr)
         {
             // Suggest dereferencing the argument to a function/method call if possible
-
             let mut real_trait_pred = trait_pred;
             while let Some((parent_code, parent_trait_pred)) = code.parent() {
                 code = parent_code;
@@ -553,6 +552,9 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                         );
                         if self.predicate_may_hold(&obligation)
                             && self.predicate_must_hold_modulo_regions(&sized_obligation)
+                            // Do not suggest * if it is already a reference,
+                            // will suggest removing the borrow instead in that case.
+                            && !matches!(expr.kind, hir::ExprKind::AddrOf(..))
                         {
                             let call_node = self.tcx.hir_node(*call_hir_id);
                             let msg = "consider dereferencing here";

--- a/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.rs
+++ b/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.rs
@@ -1,0 +1,18 @@
+fn main() {
+    let fields = vec![1];
+    let variant = vec![2];
+
+    // should not suggest `*&variant.iter()`
+    for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
+        //~^ ERROR `&std::slice::Iter<'_, {integer}>` is not an iterator
+        //~| ERROR `&std::slice::Iter<'_, {integer}>` is not an iterator
+        eprintln!("{} {}", src, dest);
+    }
+
+    // don't suggest add `variant.iter().clone().clone()`
+    for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone()) {
+        //~^ ERROR `&std::slice::Iter<'_, {integer}>` is not an iterator
+        //~| ERROR `&std::slice::Iter<'_, {integer}>` is not an iterator
+        eprintln!("{} {}", src, dest);
+    }
+}

--- a/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.stderr
+++ b/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.stderr
@@ -1,0 +1,69 @@
+error[E0277]: `&std::slice::Iter<'_, {integer}>` is not an iterator
+  --> $DIR/invalid-suggest-deref-issue-127590.rs:6:54
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
+   |                        --------------                ^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, {integer}>` is not an iterator
+   |                        |
+   |                        required by a bound introduced by this call
+   |
+   = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, {integer}>`, which is required by `&std::slice::Iter<'_, {integer}>: IntoIterator`
+   = note: required for `&std::slice::Iter<'_, {integer}>` to implement `IntoIterator`
+note: required by a bound in `std::iter::zip`
+  --> $SRC_DIR/core/src/iter/adapters/zip.rs:LL:COL
+help: consider dereferencing here
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter()) {
+   |                                                      +
+help: consider removing the leading `&`-reference
+   |
+LL -     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
+LL +     for (src, dest) in std::iter::zip(fields.iter(), variant.iter()) {
+   |
+
+error[E0277]: `&std::slice::Iter<'_, {integer}>` is not an iterator
+  --> $DIR/invalid-suggest-deref-issue-127590.rs:6:24
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, {integer}>` is not an iterator
+   |
+   = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, {integer}>`, which is required by `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>: IntoIterator`
+   = help: the trait `Iterator` is implemented for `std::slice::Iter<'a, T>`
+   = note: required for `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>` to implement `Iterator`
+   = note: required for `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>` to implement `IntoIterator`
+
+error[E0277]: `&std::slice::Iter<'_, {integer}>` is not an iterator
+  --> $DIR/invalid-suggest-deref-issue-127590.rs:13:54
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone()) {
+   |                        --------------                ^^^^^^^^^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, {integer}>` is not an iterator
+   |                        |
+   |                        required by a bound introduced by this call
+   |
+   = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, {integer}>`, which is required by `&std::slice::Iter<'_, {integer}>: IntoIterator`
+   = note: required for `&std::slice::Iter<'_, {integer}>` to implement `IntoIterator`
+note: required by a bound in `std::iter::zip`
+  --> $SRC_DIR/core/src/iter/adapters/zip.rs:LL:COL
+help: consider dereferencing here
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter().clone()) {
+   |                                                      +
+help: consider removing the leading `&`-reference
+   |
+LL -     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone()) {
+LL +     for (src, dest) in std::iter::zip(fields.iter(), variant.iter().clone()) {
+   |
+
+error[E0277]: `&std::slice::Iter<'_, {integer}>` is not an iterator
+  --> $DIR/invalid-suggest-deref-issue-127590.rs:13:24
+   |
+LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone()) {
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&std::slice::Iter<'_, {integer}>` is not an iterator
+   |
+   = help: the trait `Iterator` is not implemented for `&std::slice::Iter<'_, {integer}>`, which is required by `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>: IntoIterator`
+   = help: the trait `Iterator` is implemented for `std::slice::Iter<'a, T>`
+   = note: required for `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>` to implement `Iterator`
+   = note: required for `Zip<std::slice::Iter<'_, {integer}>, &std::slice::Iter<'_, {integer}>>` to implement `IntoIterator`
+
+error: aborting due to 4 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.stderr
+++ b/tests/ui/traits/suggest-dereferences/invalid-suggest-deref-issue-127590.stderr
@@ -10,10 +10,6 @@ LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
    = note: required for `&std::slice::Iter<'_, {integer}>` to implement `IntoIterator`
 note: required by a bound in `std::iter::zip`
   --> $SRC_DIR/core/src/iter/adapters/zip.rs:LL:COL
-help: consider dereferencing here
-   |
-LL |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter()) {
-   |                                                      +
 help: consider removing the leading `&`-reference
    |
 LL -     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter()) {
@@ -43,10 +39,6 @@ LL |     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone(
    = note: required for `&std::slice::Iter<'_, {integer}>` to implement `IntoIterator`
 note: required by a bound in `std::iter::zip`
   --> $SRC_DIR/core/src/iter/adapters/zip.rs:LL:COL
-help: consider dereferencing here
-   |
-LL |     for (src, dest) in std::iter::zip(fields.iter(), *&variant.iter().clone()) {
-   |                                                      +
 help: consider removing the leading `&`-reference
    |
 LL -     for (src, dest) in std::iter::zip(fields.iter(), &variant.iter().clone()) {

